### PR TITLE
Add various small Kotlin extensions

### PR DESCRIPTION
**Pull request by JohannesZick**
Hi,

first off, I've read the contributing guide. And no, this is not a bugfix. But the extensions are so small that I felt this is easier than explaining everything in text on [insert medium of choice here]. So, feel free to close this, but maybe consider adding some of these into the next release. I'm sure a lot of people keep writing these extensions all the time in their projects.

If you have questions or suggestions, please just ask.

### CODE REVIEW

**Comment by lukaseder** [comment]
Thanks a lot for your suggestions.

I've never thought about PRs this way, but I agree that a PR to convey some ideas is a very good approach, as long as there is no expectation for it to be merged. I've updated the contribution guidelines. Thanks for that idea.

I will comment on the individual suggestions directly in the commit.

File: jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt [hunk 1]
```diff
@@ -83,6 +87,15 @@ inline fun Field<Boolean>.not(): Condition = condition(this).not()
...
 
+@Support
+inline fun <reified T: Any> TableLike<*>.field(index: Int) = this.field(index, T::class.java)
+
+@Support
+inline fun <reified T: Any> TableLike<*>.field(name: Name) = this.field(name, T::class.java)
+
+@Support
+inline fun <reified T: Any> TableLike<*>.field(name: String) = this.field(name, T::class.java)
+
```
Comment (lukaseder) [inline jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt:98]: These types of suggestions have been made before by Jake Wharton
https://github.com/jOOQ/jOOQ/issues/6256#issuecomment-512314479

And later rejected after thorough analysis:
https://github.com/jOOQ/jOOQ/issues/6256#issuecomment-625175287

TL;DR: While looking very enticing at first, it's too late to add such convenience to the vast jOOQ API, as:

- It would be a *huge* change with many many methods involved
- Some methods could not have this kotlin-idiom applied consistently, leading to more cognitive friction than benefit

File: jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt [hunk 2]
```diff
@@ -220,3 +233,59 @@ inline fun <reified T: Any> Select<Record1<T>>.notIn(vararg values: T): Conditio
...
+inline fun <reified T: Any> Fetchable<*>.fetchOneInto() = this.fetchOneInto(T::class.java)
+
+@Support
+inline fun <reified T: Any> Fetchable<*>.fetchOptionalInto() = this.fetchOptionalInto(T::class.java)
+
+@Support
+inline fun <reified T: Any> Fetchable<*>.fetchSingleInto() = this.fetchSingleInto(T::class.java)
+
+@Support
+inline fun <reified T: Any> Fetchable<*>.fetchStreamInto() = this.fetchStreamInto(T::class.java)
```
Comment (lukaseder) [inline jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt:263]: Same as above.

File: jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt [hunk 3]
```diff
@@ -220,3 +233,59 @@ inline fun <reified T: Any> Select<Record1<T>>.notIn(vararg values: T): Conditio
...
+suspend fun <R: Record> Query.executeAwait(executor: Executor): Int = suspendCoroutine {
+        this.executeAsync(executor)
+                .handle { result, error -> if (error != null) it.resumeWithException(error) else it.resume(result) }
+}
+
+@Support
+suspend fun Query.executeAwait(): Int = suspendCoroutine {
+        this.executeAsync()
+                .handle { result, error -> if (error != null) it.resumeWithException(error) else it.resume(result) }
+}
```
Comment (lukaseder) [inline jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt:291]: These kinds of things are being discussed here: https://github.com/jOOQ/jOOQ/issues/9335

I don't feel at ease with the status quo of having:

- Classic blocking API (which will "soon" be supercharged by Loom)
- Asynchronous API via JDK `CompletionStage`
- Reactive API via reactive-streams or JDK `Flow`
- Kotlin specific coroutine API

All of jOOQ's API needs to be duplicated many times, with a lot cognitive overhead for users and maintenance overhead on our side.

I'm still betting on Loom, hoping to sit out all of the above fads, which are *much more* thoroughly addressed by Loom, without the need to change a single thing to jOOQ's API (or our internal JDBC usage).

**Comment by lukaseder** [comment]
Since all the suggestions are already addressed in other issues, I'm closing this PR as a duplicate.

Thanks again for your suggestions. If you have more ideas, just let me know. I'll be very happy to discuss them!

File: jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt (hunk 1 above)
Comment (JohannesZick) [reply jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt:98]: Ah, thanks for the link. These are the only ones I was aware of, but I really don't have a good oversight of the full API. At some point it just becomes too much, I agree.

File: jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt (hunk 3 above)
Comment (JohannesZick) [reply jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt:291]: Thanks again for the link. I'll add my two cents there.

Comment (hantsy) [reply jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt:291]: Currently the Select is a `Publisher`, it is easy to convert to Kotlin Coroutines via Reactor kotlin extensions.

Comment (lukaseder) [reply jOOQ-kotlin/src/main/kotlin/org/jooq/kotlin/Extensions.kt:291]: Can you show an example, @hantsy ?
